In [34]:
import pandas as pd
import numpy as np

## --- Day 1: Sonar Sweep ---

#### Part1:
As the submarine drops below the surface of the ocean, it automatically performs a sonar sweep of the nearby sea floor. On a small screen, the sonar sweep report (your puzzle input) appears: each line is a measurement of the sea floor depth as the sweep looks further and further away from the submarine.

This report indicates that, scanning outward from the submarine, the sonar sweep found depths of 199, 200, 208, 210, and so on.

The first order of business is to figure out how quickly the depth increases, just so you know what you're dealing with - you never know if the keys will get carried into deeper water by an ocean current or a fish or something.

To do this, count the number of times a depth measurement increases from the previous measurement. (There is no measurement before the first measurement.) In the example above, the changes are as follows:

199 (N/A - no previous measurement) <br>
200 (increased) <br>
208 (increased) <br>
210 (increased)<br>
200 (decreased)<br>
207 (increased)<br>
240 (increased)<br>
269 (increased)<br>
260 (decreased)<br>
263 (increased)<br>
In this example, there are 7 measurements that are larger than the previous measurement.

How many measurements are larger than the previous measurement?

In [20]:
with open('./example_input.txt') as f:
    ex_lines = f.readlines()

lst = [int(x.strip()) for x in ex_lines]
lst

[199, 200, 208, 210, 200, 207, 240, 269, 260, 263]

In [21]:
ex_count = 0
for i in range(1, len(lst)):
    if lst[i] > lst[i-1]:
        ex_count += 1
ex_count

7

We were able to get the correct answer (7) using the above code for the example input, lets see if it works with the real input!

In [22]:
with open('./input1.txt') as f:
    lines = f.readlines()
lst = [int(x.strip()) for x in lines]
lst

[122,
 129,
 116,
 122,
 135,
 148,
 151,
 158,
 161,
 179,
 197,
 198,
 187,
 210,
 211,
 206,
 207,
 210,
 211,
 216,
 222,
 224,
 222,
 234,
 237,
 238,
 239,
 245,
 276,
 262,
 268,
 266,
 293,
 291,
 293,
 296,
 298,
 291,
 292,
 295,
 297,
 298,
 297,
 300,
 297,
 309,
 315,
 319,
 320,
 319,
 323,
 338,
 349,
 351,
 352,
 354,
 348,
 366,
 365,
 388,
 385,
 379,
 381,
 385,
 391,
 392,
 405,
 407,
 415,
 427,
 433,
 436,
 439,
 444,
 450,
 449,
 453,
 457,
 459,
 460,
 467,
 471,
 474,
 473,
 477,
 487,
 489,
 498,
 512,
 516,
 507,
 526,
 544,
 560,
 570,
 569,
 580,
 590,
 591,
 593,
 596,
 597,
 598,
 604,
 599,
 608,
 603,
 616,
 649,
 636,
 649,
 648,
 653,
 654,
 655,
 656,
 664,
 666,
 669,
 663,
 661,
 682,
 681,
 685,
 693,
 701,
 727,
 706,
 704,
 705,
 706,
 711,
 716,
 711,
 721,
 728,
 724,
 734,
 733,
 737,
 736,
 742,
 743,
 742,
 744,
 745,
 749,
 750,
 760,
 767,
 787,
 788,
 787,
 789,
 785,
 798,
 796,
 797,
 798,
 800,
 807,
 817,
 829,
 828,
 830,
 831,
 863

In [23]:
count = 0
for i in range(1, len(lst)):
    if lst[i] > lst[i-1]:
        count += 1
count

1462

1462 was the correct answer according the AOC, time for part 2!

#### Part 2

Considering every single measurement isn't as useful as you expected: there's just too much noise in the data.

Instead, consider sums of a three-measurement sliding window. Again considering the above example:


Start by comparing the first and second three-measurement windows. The measurements in the first window are marked A (199, 200, 208); their sum is 199 + 200 + 208 = 607. The second window is marked B (200, 208, 210); its sum is 618. The sum of measurements in the second window is larger than the sum of the first, so this first comparison increased.

Your goal now is to count the number of times the sum of measurements in this sliding window increases from the previous sum. So, compare A with B, then compare B with C, then C with D, and so on. Stop when there aren't enough measurements left to create a new three-measurement sum.

In the above example, the sum of each three-measurement window is as follows:

A: 607 (N/A - no previous sum)<br>
B: 618 (increased) <br>
C: 618 (no change)<br>
D: 617 (decreased)<br>
E: 647 (increased)<br>
F: 716 (increased)<br>
G: 769 (increased)<br>
H: 792 (increased)<br>
In this example, there are 5 sums that are larger than the previous sum.

Consider sums of a three-measurement sliding window. How many sums are larger than the previous sum?



This question is similair to part 1, but with an extra step to sum adjecent trios, lets give it a shot!

In [31]:
with open('./example_input.txt') as f:
    ex_lines = f.readlines()
lst = [int(x.strip()) for x in ex_lines]

In [36]:
sums = []
for i in range(2, len(lst)):
    window = [lst[i], lst[i-1], lst[i-2]]
    Sum = sum(window)
    sums.append(Sum)
sums

[607, 618, 618, 617, 647, 716, 769, 792]

In [38]:
ex_count=0
for i in range(1, len(sums)):
    if sums[i] > sums[i-1]:
        ex_count += 1
ex_count

5

In [39]:
with open('./input1.txt') as f:
    lines = f.readlines()
lst = [int(x.strip()) for x in lines]

In [42]:
sums = []
for i in range(2, len(lst)):
    window = [lst[i], lst[i-1], lst[i-2]]
    Sum = sum(window)
    sums.append(Sum)

In [43]:
answer = 0
for i in range(1, len(sums)):
    if sums[i] > sums[i-1]:
        answer +=1
answer

1497

Final answer is 1497 sums of 3 larger than the previous sum of 3!